tests/unit/s2n_random_test.c

```diff
@@ -142,6 +146,110 @@ static S2N_RESULT s2n_basic_pattern_tests(S2N_RESULT (*s2n_get_random_data_cb)(s
+    current_bound = 1;
+    EXPECT_OK(s2n_get_range_cb(current_bound, &current_output));
+    EXPECT_EQUAL(current_output, 0);
+
+    /* For a upper bound of 2, 0 and 1 should be the only possible outputs */
+    current_bound = 1;
+    EXPECT_OK(s2n_get_range_cb(current_bound, &current_output));
+    EXPECT_TRUE((current_output == 0) || (current_output == 1));
+
+    /* Test NUMBER_OF_BOUNDS upper bounds. For each resulting range, draw
+     * NUMBER_OF_RANGE_FUNCTION_CALLS numbers from s2n_public_random() and
+     * verify the output. Set 2^30 * NUMBER_OF_RANGE_FUNCTION_CALLS as the
+     * minimal value for the upper bound.
+     */
+    int64_t minimal_upper_bound = (int64_t) 0x40000000 * (int64_t) NUMBER_OF_RANGE_FUNCTION_CALLS;
+    for (size_t bound_ctr = 0; bound_ctr < NUMBER_OF_BOUNDS; bound_ctr++) {
+
+UPPER_BOUND_TOO_SMALL:
+        EXPECT_OK(s2n_get_private_random_data(&upper_bound_blob));
+        if (chosen_upper_bound < minimal_upper_bound) {
+            goto UPPER_BOUND_TOO_SMALL;
+        }
+
+        for (size_t func_call_ctr = 0; func_call_ctr < NUMBER_OF_RANGE_FUNCTION_CALLS; func_call_ctr++) {
+            EXPECT_OK(s2n_public_random(chosen_upper_bound, &range_results[func_call_ctr]));
+            EXPECT_TRUE(range_results[func_call_ctr] < chosen_upper_bound);
+        }
+
+        /* The probability of "at least MAX_REPEATED_OUTPUT repeated values"
+         * follows a binomial distribution. Hence, we can get an upper bound via
+         * Markov's inequality:
+         * P("at least MAX_REPEATED_OUTPUT repeated values")
+         *      <= E("at least MAX_REPEATED_OUTPUT repeated values") / MAX_REPEATED_OUTPUT.
+         *       = (NUMBER_OF_RANGE_FUNCTION_CALLS * 1/(2^30 * NUMBER_OF_RANGE_FUNCTION_CALLS)) / MAX_REPEATED_OUTPUT
+         *       = 1/(2^30 * MAX_REPEATED_OUTPUT)
+         *
+         * With current parameters
+         *   NUMBER_OF_BOUNDS = 10
+         *   MAX_REPEATED_OUTPUT = 4
+         * this ends up with about a ~1/2^30 probability of failing this test
+         * with a false positive.
+         *
+         * O(n^2) sorting below, but NUMBER_OF_RANGE_FUNCTION_CALLS is very
+         * small, so no big biggie. Sorting the array means that we can check
+         * for reapeated numbers by just counting from left to right resetting
+         * the count when meeting a different value.
+         */
+        sort_array(range_results);
+        uint64_t current_value = range_results[0];
+        size_t repeat_count = 1;
+        for (size_t ctr = 1; ctr < NUMBER_OF_RANGE_FUNCTION_CALLS - 1; ctr++) {
+
+            if (current_value == range_results[ctr]) {
+                repeat_count = repeat_count + 1;
+            } else {
+                current_value = range_results[ctr];
+                repeat_count = 1;
+            }
+
+            EXPECT_TRUE(repeat_count < MAX_REPEATED_OUTPUT);
+        }
+
+        /* Reset for next iteration */
+        memset(range_results, 0, sizeof(range_results));
```

Comment:
https://github.com/aws/s2n-tls/blob/main/utils/s2n_safety_macros.h#L176

Reply:
If I just replace like this
```
RESULT_CHECKED_MEMSET(range_results, 0, sizeof(range_results));
```
I get the following error:
```
$ ninja-build
[1/2] Building C object CMakeFiles/s2n_random_test.dir/tests/unit/s2n_random_test.c.o
FAILED: CMakeFiles/s2n_random_test.dir/tests/unit/s2n_random_test.c.o 
/usr/bin/cc  -I../api -I../. -I../tests -I../ -isystem /home/ec2-user/humble/aws-lc/awslc-install/include -O2 -DNDEBUG   -Wno-implicit-function-declaration -Wno-deprecated -D_POSIX_C_SOURCE=200809L -std=gnu99 -DS2N_SIKE_P434_R3_ASM -DS2N_BIKE_R3_AVX2 -DS2N_BIKE_R3_AVX512 -DS2N_BIKE_R3_PCLMUL -DS2N_KYBER512R3_AVX2_BMI2 -DS2N_ADX -DS2N_HAVE_EXECINFO -DS2N_CPUID_AVAILABLE -DS2N_FEATURES_AVAILABLE -fPIC -DS2N_FALL_THROUGH_SUPPORTED -DS2N___RESTRICT__SUPPORTED -DS2N_MADVISE_SUPPORTED -DS2N_CLONE_SUPPORTED -DS2N_LIBCRYPTO_SUPPORTS_EVP_MD5_SHA1_HASH -DS2N_LIBCRYPTO_SUPPORTS_EVP_RC4 -DS2N_LIBCRYPTO_SUPPORTS_EVP_MD_CTX_SET_PKEY_CTX -pthread -MD -MT CMakeFiles/s2n_random_test.dir/tests/unit/s2n_random_test.c.o -MF CMakeFiles/s2n_random_test.dir/tests/unit/s2n_random_test.c.o.d -o CMakeFiles/s2n_random_test.dir/tests/unit/s2n_random_test.c.o   -c ../tests/unit/s2n_random_test.c
In file included from .././error/s2n_errno.h:21:0,
                 from ../tests/s2n_test.h:25,
                 from ../tests/unit/s2n_random_test.c:22:
../tests/unit/s2n_random_test.c: In function ‘s2n_tests_get_range’:
.././utils/s2n_ensure.h:64:31: error: invalid initializer
       __typeof( d ) __tmp_d = ( d );                                             \
                               ^
.././utils/s2n_safety_macros.h:176:64: note: in expansion of macro ‘__S2N_ENSURE_SAFE_MEMSET’
 #define RESULT_CHECKED_MEMSET(destination, value, len)         __S2N_ENSURE_SAFE_MEMSET((destination), (value), (len), RESULT_ENSURE_REF)
                                                                ^~~~~~~~~~~~~~~~~~~~~~~~
../tests/unit/s2n_random_test.c:253:9: note: in expansion of macro ‘RESULT_CHECKED_MEMSET’
         RESULT_CHECKED_MEMSET(range_results, 0, sizeof(range_results));
         ^~~~~~~~~~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
```

But the following doesn't give any errors(???)
```
RESULT_CHECKED_MEMSET(&range_results[0], 0, sizeof(range_results));
```

Reply:
I forgot the macro hates actual stack arrays -_- It assumes you're giving it a pointer to the start of an array. I think you can also satisfy it with a type cast, like `RESULT_CHECKED_MEMSET((uint_8*)range_results, 0, sizeof(range_results));`?